src/meta-srv/src/election/postgres.rs

```diff
@@ -0,0 +1,488 @@
+        loop {
+            let _ = keep_alive_interval.tick().await;
+
+            let prev = self.get_value_with_lease(&key).await?.unwrap_or_default();
+            let now = time::SystemTime::now()
+                .duration_since(time::UNIX_EPOCH)
+                .unwrap_or_default()
+                .as_secs_f64();
+
+            ensure!(
+                prev.expire_time > now,
+                UnexpectedSnafu {
+                    violated: format!(
+                        "Candidate lease expired, key: {:?}",
+                        String::from_utf8_lossy(&key.into_bytes())
+                    ),
+                }
+            );
+
+            let updated = ValueWithLease {
+                value: prev.value.clone(),
+                expire_time: now + CANDIDATE_LEASE_SECS as f64,
+            };
+            self.update_value_with_lease(&key, &prev, &updated).await?;
+        }
+    }
+
+    async fn all_candidates(&self) -> Result<Vec<MetasrvNodeInfo>> {
+        let key_prefix = self.candidate_root();
+        let mut candidates = self.get_value_with_lease_by_prefix(&key_prefix).await?;
+        let now = time::SystemTime::now()
+            .duration_since(time::UNIX_EPOCH)
+            .unwrap_or_default()
+            .as_secs_f64();
+        // Remove expired candidates
+        candidates.retain(|c| c.expire_time > now);
+        let mut valid_candidates = Vec::with_capacity(candidates.len());
+        for c in candidates {
+            let node_info: MetasrvNodeInfo =
+                serde_json::from_str(&c.value).with_context(|_| DeserializeFromJsonSnafu {
+                    input: format!("{:?}", c.value),
+                })?;
+            valid_candidates.push(node_info);
+        }
+        Ok(valid_candidates)
+    }
+
+    async fn campaign(&self) -> Result<()> {
+        todo!()
+    }
+
+    async fn leader(&self) -> Result<Self::Leader> {
+        todo!()
+    }
+
+    async fn resign(&self) -> Result<()> {
+        todo!()
+    }
+}
+
+impl PgElection {
+    async fn get_value_with_lease(&self, key: &String) -> Result<Option<ValueWithLease>> {
+        let prev = self
+            .client
+            .query(POINT_GET, &[key])
+            .await
+            .context(PostgresExecutionSnafu)?;
+
+        if let Some(row) = prev.first() {
+            let value: String = row.get(1);
+            let value_with_lease: ValueWithLease =
+                serde_json::from_str(&value).with_context(|_| DeserializeFromJsonSnafu {
+                    input: format!("{value:?}"),
+                })?;
+            Ok(Some(value_with_lease))
+        } else {
+            Ok(None)
+        }
+    }
+
+    async fn get_value_with_lease_by_prefix(
+        &self,
+        key_prefix: &String,
+    ) -> Result<Vec<ValueWithLease>> {
+        let key_prefix = format!("{}%", key_prefix);
+        let prev = self
+            .client
+            .query(PREFIX_SCAN, &[&key_prefix])
```

Comment:
Maybe we can wrap the query up with a function for type safety. It's quite dangerous here to directly use strings.